Unify fixed point & Interest Accrual

branan opened the pull request.

branan commented on pallets/loans/src/math.rs:
```diff
@@ -31,43 +30,36 @@ pub fn calculate_accumulated_rate<Rate: FixedPointNumber>(
 /// calculates the debt using debt=principal_debt * cumulative_rate
-pub fn debt<Amount: FixedPointNumber, Rate: FixedPointNumber>(
-	principal_debt: Amount,
+pub fn debt<Balance: FixedPointOperand, Rate: FixedPointNumber>(
```
This gets used in the tests a ton. Should probably get moved into the testutils module, but for now I've left it.

offerijns commented on pallets/loans/src/functions.rs:
```diff
@@ -354,43 +363,55 @@ impl<T: Config> Pallet<T> {
-				ensure!(amount.is_positive(), Error::<T>::LoanValueInvalid);
+				ensure!(amount > Zero::zero(), Error::<T>::LoanValueInvalid);
+
+				// TODO: this should calculate debt at the last NAV update
```
I think this TODO can be removed
```suggestion
```

branan commented:
@NunoAlexandre This is the PR I pinged you about last week. There are two big changes:

1) Previously, the loans pallet used a custom FixedPoint number for holding currency amounts. This sort-of-simplified some intermediate computations but required a bunch of conversation to and from the `Balance` type. We now keep currency amounts in the `Balance` type just like every other pallet.

2) Instead of storing accumulated interest rates in each loan individually, the storage is now centralized. The goal is to get to a place where we can accumulate the rates in a separate operation from using them - that's not quite done yet and our benchmarks can't yet reflect the optimization, but it's a step in the right direction. Importantly, the storage for the loans pallet has been updated for it, so we won't need to make a big migration when we finish the work.

NunoAlexandre commented on pallets/loans/src/math.rs:
```diff
@@ -31,43 +30,36 @@ pub fn calculate_accumulated_rate<Rate: FixedPointNumber>(
-pub enum Adjustment<Amount: FixedPointNumber> {
-	Inc(Amount),
-	Dec(Amount),
+pub enum Adjustment<Balance> {
+	Inc(Balance),
+	Dec(Balance),
 }
```
I would leave a comment here explaining that it's meant to represent how much some other value should be incremented or decremented by. When I first read this, I thought this was meant to be Like `Inc(x) == x+1`.

offerijns replied:
Done!

NunoAlexandre commented on pallets/loans/src/math.rs:
```diff
@@ -346,20 +290,20 @@ mod tests {
 	fn test_bullet_loan_expected_cash_flow() {
 		// debt is 100
-		let debt = Amount::from_inner(100 * USD);
+		let debt: u128 = 100 * USD;
```
```suggestion
		let debt: Balance = 100 * USD;
```

NunoAlexandre reviewed (approved):
Looks good to me, left minor comments but leave it up to you 👍

NunoAlexandre commented on pallets/loans/src/math.rs:
```diff
@@ -31,43 +30,36 @@ pub fn calculate_accumulated_rate<Rate: FixedPointNumber>(
-	})
+) -> Option<Balance> {
+	use Adjustment::*;
+	let current_debt = match adjustment {
+		Inc(amount) => current_debt.checked_add(&amount),
+		Dec(amount) => current_debt.checked_sub(&amount),
+	}?;
```
That's a good point, it definitely reads better.

offerijns replied:
Is it fine if we create a ticket for making this change in the future though @NunoAlexandre ? I don't think this should block merging this PR.

NunoAlexandre replied:
Yeah, absolutely not a blocker, that sounds good to me 👍

NunoAlexandre reviewed (approved):
💯 nice work!

branan commented:
Changes since last review:
* Resolve merge conflict with RPC-related changes
* re-sign all commits since some early ones were done before we required signing

This just needs a new :+1: from somebody and it is good to go
